pkg/executor/executor.go

```diff
@@ -410,19 +411,58 @@ type DDLJobRetriever struct {
 	activeRoles    []*auth.RoleIdentity
 	cacheJobs      []*model.Job
 	TZLoc          *time.Location
+	extractor      base.MemTablePredicateExtractor
 }
 
 func (e *DDLJobRetriever) initial(txn kv.Transaction, sess sessionctx.Context) error {
-	m := meta.NewMeta(txn)
-	jobs, err := ddl.GetAllDDLJobs(sess)
-	if err != nil {
-		return err
+	skipRunningJobs := false
+	skipHistoryJobs := false
+	schemaNames := set.NewStringSet()
+	tableNames := set.NewStringSet()
+
+	ex, ok := e.extractor.(*plannercore.InfoSchemaDDLExtractor)
+	if ok {
+		// Using state to determine whether we can skip checking running/history ddl jobs
+		if states, ok := ex.ColPredicates["state"]; ok {
+			skipHistoryJobs = true
+			skipRunningJobs = true
+			states.IterateWith(func(s string) {
+				ss := strings.ToLower(s)
+				if ss == "cancelled" || ss == "synced" {
+					skipHistoryJobs = false
+				} else {
+					skipRunningJobs = false
+				}
+			})
+		}
+
+		schemaNames = ex.ColPredicates["db_name"]
+		tableNames = ex.ColPredicates["table_name"]
 	}
-	e.historyJobIter, err = ddl.GetLastHistoryDDLJobsIterator(m)
-	if err != nil {
-		return err
+
+	var err error
+
+	if !skipRunningJobs {
+		// We cannot use table_id and schema_id to construct predicates for the tidb_ddl_job table.
+		// For instance, in the case of the SQL like `create table t(id int)`,
+		// the tableInfo for 't' will not be available in the infoschema until the job is completed.
+		// As a result, we cannot retrieve its table_id.
+		// Therefore, the only choice is to scan the entire table and reduce the number of calls to unmarshal.
+		e.runningJobs, err = ddl.GetNeededDDLJobs(sess, schemaNames, tableNames)
+		if err != nil {
+			return err
+		}
 	}
-	e.runningJobs = jobs
+
+	if !skipHistoryJobs {
+		// For the similar reason, we can only use schema_name and table_name to do filtering here.
+		m := meta.NewMeta(txn)
```

Comment:
I'm not sure we should provide a consistent snapshot of `e.runningJobs` and `e.historyJobIter` or let caller to resolve the duplication. Because a job may get finished after line 451

Reply:
If we want to query running job only, this doesn't matter?

Reply:
If the user queries both running job and finished job, it may appear twice.

Reply:
If "Remove job from tidb_ddl_job` and "Add job info into history jobs` is not atomic, this problem would appear. IMO, it's not related to this optimization? 
If we want to solve this problem, maybe we need to rewrite the query itself, for example, use `distinct`.

Reply:
Oh I see that `sess` and `txn` are at the same snapshot, there's no problem. I was thinking they are two independent variables.